### PR TITLE
Eilenberg Maclane spaces and groupoid quotients

### PR DESCRIPTION
**Pull request by felixwellen**
This is a fork from Ulrik Buchholtz' PR #394, which tries to adapt to the changes on master, but type checking does not terminate in Cubical/Algebra/Groups/EilenbergMacLane1.agda
(when REComp is activated)

### CODE REVIEW

File: Cubical/HITs/EilenbergMacLane1/Base.agda [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  open GroupStr str
+
+  data EM₁ : Type ℓ where
+    embase : EM₁
+    emloop : G → embase ≡ embase
+    emcomp : (g h : G)
+           → PathP (λ j → embase ≡ emloop h j) (emloop g) (emloop (g + h))
+    emsquash : ∀ (x y : EM₁) (p q : x ≡ y) (r s : p ≡ q) → r ≡ s
+
+  {- The comp// constructor fills the square:
```
Comment (ecavallo) [inline Cubical/HITs/EilenbergMacLane1/Base.agda:27]: `comp//` -> `emcomp`?

File: Cubical/HITs/EilenbergMacLane1/Properties.agda [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+  elimEq : {B : EM₁ (G , str) → Type ℓ}
+           (Bprop : (x : EM₁ (G , str)) → isProp (B x))
+           {x y : EM₁ (G , str)}
+           (eq : x ≡ y)
+           (bx : B x)
+           (by : B y) →
+           PathP (λ i → B (eq i)) bx by
+  elimEq {B = B} Bprop {x = x} =
+    J (λ y eq → ∀ bx by → PathP (λ i → B (eq i)) bx by) (λ bx by → Bprop x bx by)
```
Comment (ecavallo) [inline Cubical/HITs/EilenbergMacLane1/Properties.agda:42]: This is an instance of a general lemma:
```
  elimEq Bprop eq = isProp→PathP (λ i → Bprop (eq i))
```
Is it worth it to give this special case a name? I noticed now that a similar lemma is defined in `SetQuotients.Properties` and `GroupoidQuotients.Properties`, but it doesn't look like they're used except to prove the `elimProp` equivalent...

File: Cubical/HITs/EilenbergMacLane1/Properties.agda [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+  elimProp2 : {C : EM₁ (G , str) → EM₁ (G , str) → Type ℓ}
+            → ((x y : EM₁ (G , str)) → isProp (C x y))
+            → C embase embase
+            → (x y : EM₁ (G , str))
+            → C x y
+  elimProp2 Cprop c = elimProp (λ x → isPropΠ (λ y → Cprop x y))
+                               (elimProp (λ y → Cprop embase y) c)
+
+  elimSet : {B : EM₁ (G , str) → Type ℓ}
```
Comment (ecavallo) [inline Cubical/HITs/EilenbergMacLane1/Properties.agda:72]: Could eliminate some repetition here and in `GroupoidQuotients` by defining `elimSet` first and then using it to help define `elimProp`.

File: Cubical/HITs/GroupoidQuotients/Properties.agda [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+elimProp2 : (Rt : BinaryRelation.isTrans R)
+          → {C : A // Rt → A // Rt → Type ℓ}
+         → ((x y : A // Rt) → isProp (C x y))
+         → ((a b : A) → C [ a ] [ b ])
+         → (x y : A // Rt)
+         → C x y
+elimProp2 Rt Cprop f = elimProp Rt (λ x → isPropΠ (λ y → Cprop x y))
+                                   (λ x → elimProp Rt (λ y → Cprop [ x ] y) (f x))
+
+[]surjective : (Rt : BinaryRelation.isTrans R)
```
Comment (ecavallo) [inline Cubical/HITs/GroupoidQuotients/Properties.agda:76]: State the type in terms of `isSurjection` from `Cubical.Functions.Surjection` and rename to `isSurjection[]`?

File: Cubical/HITs/GroupoidQuotients/Properties.agda [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    → isGroupoid B
+    → (f : A → B)
+    → (feq : {a b : A} (r : R a b) → f a ≡ f b)
+    → ({a b c : A} (r : R a b) (s : R b c)
+          → Square (feq r) (feq (Rt a b c r s)) refl (feq s))
+    → (x : A // Rt)
+    → B
+rec Rt Bgpd = elim Rt (λ _ → Bgpd)
+
+module BinarySetRelation {ℓA ℓR : Level} {A : Type ℓA} (R : Rel A A ℓR) where
```
Comment (ecavallo) [inline Cubical/HITs/GroupoidQuotients/Properties.agda:129]: Move this to `Relation.Binary.Base`, next to `isPropValued`?

**Review by ecavallo — CHANGES_REQUESTED** [review]

File: Cubical/Algebra/Group/EilenbergMacLane1.agda [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+  EM₁Connected = ∣ embase ∣ , h
+    where
+      h : (y : hLevelTrunc 2 (EM₁ Ĝ)) → ∣ embase ∣ ≡ y
+      h = trElim (λ y → isOfHLevelSuc 1 (isOfHLevelTrunc 2 ∣ embase ∣ y))
+            (elimProp Ĝ (λ x → isOfHLevelTrunc 2 ∣ embase ∣ ∣ x ∣) refl)
+
+  {- since we write composition in diagrammatic order,
+     and function composition in the other order,
+     we need right multiplication here -}
+  rightEquiv : (g : G) → G ≃ G {- equivEq _ _ (funExt (λ x → (assoc x g h) ⁻¹)) -}
```
[the remark this answers is not in the document]
Comment (felixwellen) [reply Cubical/Algebra/Group/EilenbergMacLane1.agda:70]: I guess it is something I forgot to remove, so I'll do that now.

File: Cubical/HITs/EilenbergMacLane1/Base.agda (hunk 1 above)
Comment (felixwellen) [reply Cubical/HITs/EilenbergMacLane1/Base.agda:27]: Right...

File: Cubical/HITs/EilenbergMacLane1/Properties.agda (hunk 3 above)
Comment (felixwellen) [reply Cubical/HITs/EilenbergMacLane1/Properties.agda]: Yes, that would be good. Can you give me a hint how to fill that hole? (I guess you just know what to do)

Comment (felixwellen) [reply Cubical/HITs/EilenbergMacLane1/Properties.agda:44]: Solved in meeting... (for EM-Spaces)

File: Cubical/HITs/EilenbergMacLane1/Properties.agda (hunk 2 above)
Comment (felixwellen) [reply Cubical/HITs/EilenbergMacLane1/Properties.agda]: Inlined everywhere now..

File: Cubical/HITs/EilenbergMacLane1/Properties.agda (hunk 3 above)
Comment (felixwellen) [reply Cubical/HITs/EilenbergMacLane1/Properties.agda:34]: Done

File: Cubical/HITs/GroupoidQuotients/Properties.agda (hunk 4 above)
Comment (felixwellen) [reply Cubical/HITs/GroupoidQuotients/Properties.agda]: Done

File: Cubical/HITs/GroupoidQuotients/Properties.agda (hunk 5 above)
Comment (felixwellen) [reply Cubical/HITs/GroupoidQuotients/Properties.agda]: Done

File: Cubical/HITs/EilenbergMacLane1/Properties.agda (hunk 2 above)
Comment (felixwellen) [reply Cubical/HITs/EilenbergMacLane1/Properties.agda]: Done

**Comment by ice1000** [comment]
FWIW there are some commits not recognized by github. Maybe you should update your email configuration.

image

**Comment by felixwellen** [comment]
Yes, thanks!

**Pull request merged**